File: frontend/src/api/items.ts
import { api } from './client'

export interface Item {
  id: string
  name: string
  sku: string
  description?: string
  created_at: string
  updated_at: string
}

export const itemsApi = {
  list: () => api.get<Item[]>('/items').then(r => r.data),
  get: (id: string) => api.get<Item>(`/items/${id}`).then(r => r.data),
  create: (data: { name: string; sku: string; description?: string }) =>
    api.post<Item>('/items', data).then(r => r.data),
  update: (id: string, data: Partial<Pick<Item, 'name' | 'sku' | 'description'>>) =>
    api.put<Item>(`/items/${id}`, data).then(r => r.data),
}

File: backend/routers/items.py
import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from database import get_db
from models.item import Item
from schemas.item import ItemCreate, ItemUpdate, ItemResponse

router = APIRouter(prefix="/api/items", tags=["items"])


@router.get("", response_model=list[ItemResponse])
async def list_items(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Item).order_by(Item.created_at.desc()))
    return result.scalars().all()


@router.get("/{item_id}", response_model=ItemResponse)
async def get_item(item_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    item = await db.get(Item, item_id)
    if not item:
        raise HTTPException(404, "Item not found")
    return item


@router.post("", response_model=ItemResponse, status_code=201)
async def create_item(data: ItemCreate, db: AsyncSession = Depends(get_db)):
    if len(data.name) > 50:
        raise HTTPException(422, "Item name must not exceed 50 characters")
    item = Item(**data.model_dump())
    db.add(item)
    await db.commit()
    await db.refresh(item)
    return item


@router.put("/{item_id}", response_model=ItemResponse)
async def update_item(
    item_id: uuid.UUID, data: ItemUpdate, db: AsyncSession = Depends(get_db)
):
    item = await db.get(Item, item_id)
    if not item:
        raise HTTPException(404, "Item not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        if field == "name" and value and len(value) > 50:
            raise HTTPException(422, "Item name must not exceed 50 characters")
        setattr(item, field, value)
    await db.commit()
    await db.refresh(item)
    return item
